### PR TITLE
[VPC Mgr] Fix VPC Creation Performance Issue

### PR DESCRIPTION
**Pull request by kevin-zhonghao**
**Bug Description**
The creation of vpc will take a long time (avg about 0.8s), and the time will increase as the amount of data in the database increases.

**Proposed Solution**
We found that the process of creating vxlan took a lot of time, so I replaced repo.findAllItems() in addVxlanEntity() method with repo.findItem() and set up partition as the key of the map in cache.

### CODE REVIEW

**Comment by xieus** [comment]
@kevin-zhonghao In this PR description, can you detail the issue and how you fix it?

Some recommended format like the following, which was how I edited the previous PR so I would like you to practice it as well :-)

__Bug Description__
XXX

__Proposed Solution__
YYY

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/controller/SegmentController.java [hunk 1]
```diff
@@ -97,15 +102,18 @@ public SegmentWebResponseJson createSegment(@PathVariable String projectid, @Req
...
 
+            String vpcId = segmentWebRequestObject.getVpcId();
+            VpcEntity vpcState = this.vpcDatabaseService.getByVpcId(vpcId);
+
             // verify network type
             String networkType = segmentWebRequestObject.getNetworkType();
             Long key = null;
             if (NetworkTypeEnum.VXLAN.getNetworkType().equals(networkType)) {
-                key = segmentService.addVxlanEntity(networkTypeId, networkType, segmentWebRequestObject.getVpcId());
+                key = segmentService.addVxlanEntity(networkTypeId, networkType, vpcId, vpcState.getMtu());
```
[the remark this answers is not in the document]
Comment (kevin-zhonghao) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/controller/SegmentController.java:112]: Sure

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/controller/VpcController.java [hunk 2]
```diff
@@ -136,12 +136,12 @@ public VpcWebJson createVpcState(@PathVariable String projectid, @RequestBody Vp
             RestPreconditionsUtil.verifyResourceNotNull(inVpcState);
             RestPreconditionsUtil.populateResourceProjectId(inVpcState, projectid);
 
-            this.vpcDatabaseService.addVpc(inVpcState);
-
-            inVpcState = this.vpcDatabaseService.getByVpcId(inVpcState.getId());
-            if (inVpcState == null) {
-                throw new ResourcePersistenceException();
-            }
+//            this.vpcDatabaseService.addVpc(inVpcState);
```
[the remark this answers is not in the document]
Comment (kevin-zhonghao) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/controller/VpcController.java:139]: okay

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java [hunk 3]
```diff
@@ -104,43 +100,48 @@ public Long addVlanEntity(String vlanId, String networkType, String vpcId) throw
      */
     @Override
     @DurationStatistics
-    public Long addVxlanEntity(String vxlanId, String networkType, String vpcId) throws Exception {
+    public Long addVxlanEntity(String vxlanId, String networkType, String vpcId, Integer mtu) throws Exception {
 
         Long key = null;
-        String rangeId = null;
+        String p = null;
```
[the remark this answers is not in the document]
Comment (kevin-zhonghao) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java:106]: np

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java [hunk 4]
```diff
@@ -104,43 +100,48 @@ public Long addVlanEntity(String vlanId, String networkType, String vpcId) throw
...
         Random ran = new Random();
-        Map<Integer, String> partitionsAndRangeIds = new HashMap<>();
+//        Map<Integer, String> partitionsAndRangeIds = new HashMap<>();
 
         try {
             NetworkVxlanType vxlan = new NetworkVxlanType();
+            long start = System.currentTimeMillis();
```
[the remark this answers is not in the document]
Comment (kevin-zhonghao) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java:112]: Yes

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java [hunk 5]
```diff
@@ -150,7 +151,7 @@ public Long addVxlanEntity(String vxlanId, String networkType, String vpcId) thr
 
         } catch (Exception e) {
             this.vxlanRepository.deleteItem(vxlanId);
-            this.vxlanRangeRepository.releaseVxlanKey(rangeId,key);
+            this.vxlanRangeRepository.releaseVxlanKey(p,key);
```
[the remark this answers is not in the document]
Comment (kevin-zhonghao) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java:154]: np

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/controller/SegmentController.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (xieus) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/controller/SegmentController.java:112]: @kevin-zhonghao Maybe the previous comment was unclear. It meant that using a variable to replace vpcState.getMtu().

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/controller/VpcController.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (xieus) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/controller/VpcController.java:139]: @kevin-zhonghao I think you didn't check in your local commit :-) This comment is unresolved.

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (xieus) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java:106]: Same here.

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/dao/GreRangeRepository.java [hunk 6]
```diff
@@ -121,7 +122,9 @@ public synchronized Long allocateGreKey (String rangeId) throws Exception {
             }
 
             key = networkGRERange.allocateKey();
-            cache.put(networkGRERange.getId(), networkGRERange);
+            if (!key.equals(ConstantsConfig.keyNotEnoughReturnValue)) {
```
Comment (xieus) [inline services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/dao/GreRangeRepository.java:125]: Recommend to put this check into a util method which could be used for other extended validation, not only checking keyNotEnoughReturnValue.

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java [hunk 7]
```diff
@@ -104,43 +107,46 @@ public Long addVlanEntity(String vlanId, String networkType, String vpcId) throw
...
     @DurationStatistics
-    public Long addVxlanEntity(String vxlanId, String networkType, String vpcId) throws Exception {
+    public Long addVxlanEntity(String vxlanId, String networkType, String vpcId, Integer mtu) throws Exception {
 
         Long key = null;
-        String rangeId = null;
+        String partitionStringFormat = null;
         Random ran = new Random();
-        Map<Integer, String> partitionsAndRangeIds = new HashMap<>();
+//        Map<Integer, String> partitionsAndRangeIds = new HashMap<>();
```
Comment (xieus) [inline services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java:115]: Remove the comment line.

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java [hunk 8]
```diff
@@ -104,43 +107,46 @@ public Long addVlanEntity(String vlanId, String networkType, String vpcId) throw
...
 
         try {
             NetworkVxlanType vxlan = new NetworkVxlanType();
+            // the number of randomly allocating partition in vxlan range
+            int round = 1;
+
+            while (round <= ConstantsConfig.MAX_ROUNDS) {
+                // Randomly allocate a partition and check if the partition exist in db
+                int partition = ran.nextInt(NetworkType.VXLAN_PARTITION);
```
Comment (xieus) [inline services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java:124]: will the value of "partition" get changed for every round?

Also please use String method to convert to a String object.

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java [hunk 9]
```diff
@@ -104,43 +107,46 @@ public Long addVlanEntity(String vlanId, String networkType, String vpcId) throw
...
+                partitionStringFormat = partition + "";
+                NetworkVxlanRange networkVxlanRange = this.vxlanRangeRepository.findItem(partitionStringFormat);
+                if (networkVxlanRange == null) {
+//                  rangeId = UUID.randomUUID().toString();
```
Comment (xieus) [inline services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java:128]: Remove comment code.

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java [hunk 10]
```diff
@@ -104,43 +107,46 @@ public Long addVlanEntity(String vlanId, String networkType, String vpcId) throw
...
+                    int firstKey = partition * NetworkType.VXLAN_ONE_PARTITION_SIZE;
+                    int lastKey = (partition + 1) * NetworkType.VXLAN_ONE_PARTITION_SIZE;
+                    NetworkRangeRequest request = new NetworkRangeRequest(partitionStringFormat, networkType, partition, firstKey, lastKey);
+                    this.vxlanRangeRepository.createRange(request);
```
Comment (xieus) [inline services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java:132]: Recommended to return id if a creation is successful.

Something like：
```
partitionStringFormat = this.vxlanRangeRepository.createRange(request);
```

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java [hunk 11]
```diff
@@ -104,43 +107,46 @@ public Long addVlanEntity(String vlanId, String networkType, String vpcId) throw
...
+                round ++;
             }
 
-            key = this.vxlanRangeRepository.allocateVxlanKey(rangeId);
-
-            if (vpcId != null) {
-                VpcEntity vpc = this.vpcDatabaseService.getByVpcId(vpcId);
-                Integer mtu = vpc.getMtu();
-                vxlan.setMtu(mtu);
+            if (key.equals(ConstantsConfig.keyNotEnoughReturnValue)) {
```
Comment (xieus) [inline services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java:145]: 👍

File: services/vpc_manager/src/test/java/com/futurewei/alcor/vpcmanager/SegmentControllerTests.java [hunk 12]
```diff
@@ -46,6 +48,9 @@
     @MockBean
     private SegmentDatabaseService segmentDatabaseService;
 
+    @MockBean
+    private VpcDatabaseService vpcDatabaseService;
```
Comment (xieus) [inline services/vpc_manager/src/test/java/com/futurewei/alcor/vpcmanager/SegmentControllerTests.java:52]: I think we will need one more test case to validate that if all ranges are allocated and the new codes could throw out the keyNotEnough exception.

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/allocator/NetworkKeyAllocator.java [hunk 13]
```diff
@@ -25,8 +26,11 @@ public NetworkKeyAllocator(int firstKey, int lastKey) {
     @Override
     public Long allocate() throws Exception {
         int freeBit = bitSet.nextClearBit(0);
+//        if (freeBit < 0 || freeBit >= keyNum) {
```
[the remark this answers is not in the document]
Comment (kevin-zhonghao) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/allocator/NetworkKeyAllocator.java:29]: okay

File: services/vpc_manager/src/test/java/com/futurewei/alcor/vpcmanager/SegmentControllerTests.java (hunk 12 above)
Comment (kevin-zhonghao) [reply services/vpc_manager/src/test/java/com/futurewei/alcor/vpcmanager/SegmentControllerTests.java:52]: got it

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java (hunk 9 above)
Comment (kevin-zhonghao) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java:128]: Sure

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java (hunk 8 above)
Comment (kevin-zhonghao) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java:124]: Yes，I could use String.valueOf()

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/allocator/NetworkKeyAllocator.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (kevin-zhonghao) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/allocator/NetworkKeyAllocator.java]: sure

**Comment by xieus** [comment]
@kevin-zhonghao This PR has some conflicting files with your other PR #353 that just got merged to master. Should be a minor fix. Pls take a look.

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java (hunk 10 above)
Comment (kevin-zhonghao) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/service/Impl/SegmentServiceImpl.java:132]: Sure

File: services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/dao/GreRangeRepository.java (hunk 6 above)
Comment (kevin-zhonghao) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/dao/GreRangeRepository.java:125]: Yeah that's right ~ but we could make this change when we have other extended validation later, what do you think

Comment (xieus) [reply services/vpc_manager/src/main/java/com/futurewei/alcor/vpcmanager/dao/GreRangeRepository.java:125]: Sure.

**Pull request merged**